mcs/class/corlib/Test/System/StringTest.cs

```diff
@@ -92,19 +96,36 @@ public void Constructor4_Count_Negative ()
 		}
 	}
 
+	// Several tests in this file, to run properly, allocate 4GB objects.
+	// Obviously this creates problems on several kinds of systems, so we
+	// conservatively skip these tests unless we find a high-RAM environment.
+	// Checking RAM requires PerformanceCounter which is absent on mobile,
+	// so any test that calls this must be category MobileNotWorking.
+	static bool LowMemoryTestEnvironment ()
+	{
+#if MOBILE
+		return true;
+#else
+		if (!Environment.Is64BitProcess)
+			return true;
+
+		// Require 6 GB physical RAM, for the 4GB string plus 2GB headroom 
+		var pc = new PerformanceCounter ("Mono Memory", "Total Physical Memory");
+
+		return pc.RawValue < 6L*1024L*1024L*1024L;
+#endif
+	}
+
 	[Test] // ctor (Char, Int32)
+	[Category ("MobileNotWorking")]
 	public void Constructor4_LargeString ()
 	{
-		try {
-			var x = new String ('A', int.MaxValue);
-			if (Environment.Is64BitProcess) {
-				Assert.AreEqual ('A', x[0]);
-				Assert.AreEqual ('A', x[int.MaxValue - 1]);
-			}
-			else
-				Assert.Fail ("Expected OutOfMemoryException.");
-		} catch (OutOfMemoryException) {
-		}
+		if (LowMemoryTestEnvironment())
+			return;
```

Comment:
Do Assert.Ignore ("Low memory device, skipping.") here and in the other tests so we don't get a "passed" test result even though it didn't run.